torch/_dynamo/variables/builder.py

```diff
@@ -2237,45 +2244,87 @@ def _automatic_dynamic(
-                    e.ndim,
-                    frame_state_entry.size,
-                )
-                frame_state_entry.size = None
-            else:
-                # If there is already an entry, and the dim matches, for every size in the frame state which
-                # disagrees with the current static size, replace it with None. E.g., {"x": [2, 3]} -> {"x": [2, None]}
-                for i, dim in enumerate(frame_state_entry.size):
-                    if dim is not None and e.size()[i] != dim:
-                        log.debug(
-                            "automatic dynamic %s size(%s) %s != %s",
-                            name,
-                            i,
-                            e.size(i),
-                            dim,
-                        )
-                        frame_state_entry.size[i] = None
+    def update_frame_state(size, stride):
+        frame_state_entry = None
+        if name not in tx.output.frame_state:
+            # If there is no entry for this source, add the tensor to frame state with its current static size.
+            # E.g., {} -> {"x": [2, 4]}
+            frame_state_entry = FrameStateSizeEntry(None, None, None)
+            frame_state_entry.size = list(size)
+            frame_state_entry.stride = list(stride)
+        else:
+            frame_state_entry = tx.output.frame_state[name]
+            if frame_state_entry.size is not None:
+                if e.ndim != len(frame_state_entry.size):
+                    # If there is already an entry, and the dim mismatches, replace the frame state entry with None.
+                    # E.g. {"x": [2, 3, 4]} -> {"x": None}
+                    log.debug(
+                        "automatic dynamic %s dim %s != %s",
+                        name,
+                        e.ndim,
+                        frame_state_entry.size,
+                    )
+                    frame_state_entry.size = None
+                    frame_state_entry.stride = None
+                else:
+                    # If there is already an entry, and the dim matches, for every size/stride in the frame state which
+                    # disagrees with the current static size/stride, replace it with None.
+                    # E.g., {"x": [2, 3]} -> {"x": [2, # None]}
+
+                    has_size_changed = False
+                    for i, dim in enumerate(frame_state_entry.size):
+                        if dim is not None and size[i] != dim:
+                            log.debug(
+                                "automatic dynamic %s size(%s) %s != %s",
+                                name,
+                                i,
+                                e.size(i),
+                                dim,
+                            )
+                            has_size_changed = True
+                            frame_state_entry.size[i] = None
+                        elif dim is None:
+                            has_size_changed = True
```

Comment:
This is sort of overly complicated, because `has_size_changed = frame_entry.size[i] is None` after the update ;)